### PR TITLE
feat: Custom schema for SQL query visualization

### PR DESCRIPTION
**Pull request by sanjibansg**
This PR adds support for defining custom schema in the SQL CodeView. Previously, querying only on the `lineitem` table was allowed through SQL in the fiddle, but with this feature, different table schemas can be defined and queries can be visualized.

### CODE REVIEW

File: api/app.py [hunk 1]
```diff
@@ -1,20 +1,64 @@
...
+    finally:
+        conn.close()
+
+
+oauth2_scheme = OAuth2PasswordBearer(tokenUrl=allowed_urls[3])
+
+
+def verify_token(token: str = Depends(oauth2_scheme)):
+    try:
+        payload = jwt.decode(token, "key", algorithms=["HS256"])
```
Comment (sanjibansg) [inline api/app.py:54]: this shall be replaced by a ENV variable

File: client/src/assets/js/shared.js [hunk 2]
```diff
@@ -48,4 +49,15 @@ function plot(plan, status_func) {
   }
 }
 
-export { readFile, readText, validate, plot, clearGraph };
+async function generateToken(user_id) {
+  const payload = {
+    user_id: user_id,
+  };
+  const secret = new TextEncoder().encode("key");
```
Comment (sanjibansg) [inline client/src/assets/js/shared.js:56]: will be present in an ENV variable

File: client/src/components/SqlSchema.vue [hunk 3]
```diff
@@ -31,14 +31,72 @@
...
-              cols="40"
-            >
-            </textarea>
+            <div class="accordion" id="schemaAccordion">
+              <div
+                v-for="(schema, i) in schemas"
+                :key="schema.name"
+                class="accordion-item"
+              >
+                <h2 class="accordion-header" :id="`heading${schema.name}`">
```
Comment (sanjibansg) [inline client/src/components/SqlSchema.vue:40]: need to add test for the feature, and also fix the merge conflicts.

File: api/app.py [hunk 4]
```diff
@@ -1,20 +1,64 @@
...
+
 from loguru import logger
 
+
 app = FastAPI()
-con = None
+
+allowed_urls = [
+    "substrait-fiddle.com",
+    "substrait-fiddle.com",
```
Comment (EpsilonPrime) [inline api/app.py:31]: This is the same as the line before, isn't it?

File: api/app.py [hunk 5]
```diff
@@ -23,9 +67,13 @@ def ping():
 
 
 @app.on_event("startup")
-def Initialize():
-    global con
-    con = ConnectDuckDB()
+async def Initialize():
```
Comment (richtia) [inline api/app.py:70]: Python function names are generally all lowercase with words separated by an underscore.  I think you follow that elsewhere as well.  Let's stick to that.

Classes are generally camelcase with the first letter of each word as a capital letter.

File: api/backend/duckdb.py [hunk 6]
```diff
@@ -3,34 +3,49 @@
...
+
+    def check_pool(self):
+        if len(self.conn_pool) == 0:
+            for i in range(5):
+                conn = duckdb.connect("duck.db")
+                conn.install_extension("substrait")
+                conn.load_extension("substrait")
+                self.conn_pool.append(conn)
+
+    def initialize(self):
```
Comment (richtia) [inline api/backend/duckdb.py:45]: does a name like get_connection make more sense here?

File: api/backend/duckdb.py [hunk 7]
```diff
@@ -3,34 +3,49 @@
...
+        conn = duckdb.connect("duck.db")
+        conn.execute(query=query_lineitem)
+
+        for i in range(5):
+            conn = duckdb.connect("duck.db")
+            conn.install_extension("substrait")
+            conn.load_extension("substrait")
+            self.conn_pool.append(conn)
+
+    def check_pool(self):
```
Comment (richtia) [inline api/backend/duckdb.py:37]: why is this function here if we're doing something similar in the __init__?

**Comment by richtia** [comment]
Can we add docstrings to some of python functions to get a better understanding of what they're doing?

File: api/app.py [hunk 8]
```diff
@@ -1,20 +1,64 @@
...
-    ConnectDuckDB,
+    DuckDBConnection,
     CheckDuckDBConnection,
     ExecuteDuckDb,
     ParseFromDuckDB,
+    DeleteTableFromDuckDB,
 )
 
+from backend.ttl_cache import TTL_Cache
+
```
[the remark this answers is not in the document]
Comment (sanjibansg) [reply api/app.py]: made the change, thanks for the suggestion!

File: api/app.py (hunk 5 above)
Comment (sanjibansg) [reply api/app.py]: yes, that's right, I've made the changes, thanks for the suggestion!

File: api/backend/duckdb.py (hunk 6 above)
Comment (sanjibansg) [reply api/backend/duckdb.py]: yep, that's right, made the change, thanks!

File: api/app.py [hunk 9]
```diff
@@ -23,9 +67,13 @@ def ping():
...
+async def Initialize():
+    app.state.duck_pool = DuckDBConnection()
+    app.state.schema_cache = TTL_Cache(
+        maxsize=100,
+        ttl=3600,
+        on_expire=lambda key, _: DeleteTableFromDuckDB(key, get_duck_conn()),
+    )
 
 
 @app.get("/health/duckcb/")
```
[the remark this answers is not in the document]
Comment (sanjibansg) [reply api/app.py]: oh yes, sorry, I missed that, thanks for pointing that out!

File: api/app.py (hunk 4 above)
Comment (sanjibansg) [reply api/app.py]: yep, sorry, repetition error, fixed that.

File: api/backend/duckdb.py [hunk 10]
```diff
@@ -3,34 +3,49 @@
...
+                        l_shipdate DATE NOT NULL, 
+                        l_commitdate DATE NOT NULL, 
+                        l_receiptdate DATE NOT NULL, 
+                        l_shipinstruct VARCHAR NOT NULL, 
+                        l_shipmode VARCHAR NOT NULL, 
+                        l_comment VARCHAR NOT NULL);"""
+        conn = duckdb.connect("duck.db")
+        conn.execute(query=query_lineitem)
+
+        for i in range(5):
```
[the remark this answers is not in the document]
Comment (sanjibansg) [reply api/backend/duckdb.py]: agreed, thanks for the suggestion, made the change.

File: client/src/assets/js/substrait-parser.js [hunk 11]
```diff
@@ -755,7 +755,7 @@ class SubstraitParser {
         nodes.push(this.rootRelToNode(relation.root));
       } else {
         throw new Error(
-          `Unrecognized plan relation ${JSON.stringify(relation)}`
+          `Unrecognized plan relation ${JSON.stringify(relation)}`,
```
[the remark this answers is not in the document]
Comment (sanjibansg) [reply client/src/assets/js/substrait-parser.js]: ah yes, you're right, removed it, thanks!

File: client/src/App.vue [hunk 12]
```diff
@@ -110,9 +110,19 @@ import SubstraitGraph from "@/components/SubstraitGraph.vue";
...
+import { v4 as uuidv4 } from "uuid";
+
 import { clearGraph } from "./assets/js/substrait-d3.js";
+import { store } from "./components/store";
+import { generateToken } from "./assets/js/shared";
 
 export default {
+  async mounted() {
+    const sessionToken = uuidv4();
+    store.set_user(sessionToken.replace(/-/g, "_"));
```
[the remark this answers is not in the document]
Comment (sanjibansg) [reply client/src/App.vue]: yes, that sounds right, thanks for the suggestion, made the change!

**Comment by sanjibansg** [comment]
> Can we add docstrings to some of python functions to get a better understanding of what they're doing?

Yes, considering the big change, docstrings will be good for giving a better understanding of the workflow. I will add them.

File: api/test.py [hunk 13]
```diff
@@ -15,44 +16,69 @@ def test_ping():
...
+        assert response.status_code == 200
 
 
 def test_validate_binary():
-    url = "https://github.com/westonpace/substrait-viewer/blob/main/demo/q1.bin"
-    response = requests.get(url)
-    file_content = response.content
+    with TestClient(app) as client:
+        url = (
+            "https://github.com/westonpace/"
```
Comment (EpsilonPrime) [inline api/test.py:36]: This should be updated.

**Review by EpsilonPrime — APPROVED** [review]

File: client/src/components/SqlSchema.vue (hunk 3 above)
Comment (sanjibansg) [reply client/src/components/SqlSchema.vue:40]: made the change.

File: api/app.py [hunk 14]
```diff
@@ -1,46 +1,87 @@
...
-@router.post("/validate/")
-async def Validate(plan: dict, override_levels: list[int]):
+
+############################################################
+# API Endpoint to validate a substrait plan
+############################################################
+@router.post("/validate/", status_code=status.HTTP_200_OK)
+async def validate(plan: dict, override_levels: list[int]):
+    '''
+        Validates a plan using the substrait-validator
```
[the remark this answers is not in the document]
Comment (sanjibansg) [reply api/app.py]: made the change, thanks!

File: api/components/duckdb.py [hunk 15]
```diff
@@ -0,0 +1,130 @@
+import duckdb
+from fastapi import HTTPException
+from loguru import logger
+
+# Pool size is default at 5 for maintaining 
+# 5 concurrent DuckDB connection objects
+POOL_SIZE = 5
```
Comment (sanjibansg) [inline api/components/duckdb.py:7]: I think we should also make this as an ENV variable

**Pull request merged**